server/api/account.js

```diff
@@ -604,8 +632,27 @@ module.exports.init = function(server) {
 
         res.send({ success: true, message: 'Block list entry removed successfully', username: lowerCaseUser, user: updatedUser.getWireSafeDetails() });
     }));
+
+    server.post('/api/account/:username/updateavatar', passport.authenticate('jwt', { session: false }), wrapAsync(async (req, res) => {
+        let user = await checkAuth(req, res);
+
+        if(!user) {
+            return;
+        }
+
+        user = user.getDetails();
+
+        await downloadAvatar(user);
+
+        res.send({ success: true });
+    }));
 };
 
+async function downloadAvatar(user) {
+    let avatar = await util.httpRequest(`https://www.gravatar.com/avatar/${user.emailHash}?d=identicon&s=24`, { encoding: null });
+    await writeFile(`public/img/avatar/${user.username}.png`, avatar, 'binary');
```

Comment:
Is our username validation in a state such that the username will always be a valid filename? E.g. if slashes are allowed in the username ("Guns/Roses") then it will look for things under a non-existent sub directory.

Reply:
usernames must match /^[A-Za-z0-9_-]+$/ - so all good there